sound/soc/sof/control.c

```diff
@@ -180,18 +191,28 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
 		(struct soc_bytes_ext *)kcontrol->private_value;
 	struct snd_sof_control *scontrol = be->dobj.private;
 	struct snd_sof_dev *sdev = scontrol->sdev;
-	//struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	//unsigned int i, channels = scontrol->num_channels;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct sof_abi_hdr *data = cdata->data;
+	int ret = 0;
 
 	pm_runtime_get_sync(sdev->dev);
 
-	/* TODO: copy from userspace */
+	if (data->size > be->max) {
+		dev_err(sdev->dev, "error: size too big %d bytes max is %d\n",
+			data->size, be->max);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* copy from kcontrol */
+	memcpy(data, ucontrol->value.bytes.data, data->size);
```

Comment:
still looks asymmetrical between put() and get().

put: memcpy(data, ucontrol->value.bytes.data, data->size);
get: memcpy(ucontrol->value.bytes.data, data, data->size + sizeof(*data);